metrics/multilabelmap.go

```diff
@@ -0,0 +1,245 @@
+
+	m sync.Map // map[T]expvar.Var
+
+	mu     sync.RWMutex
+	sorted []labelsAndValue[T] // by labels string, to match expvar.Map + for aesthetics in output
+}
+
+// NewMultiLabelMap creates and publishes (via expvar.Publish) a new
+// MultiLabelMap[T] variable with the given name and returns it.
+func NewMultiLabelMap[T comparable](name string, promType, helpText string) *MultiLabelMap[T] {
+	m := &MultiLabelMap[T]{
+		Type: promType,
+		Help: helpText,
+	}
+	expvar.Publish(name, m)
+	return m
+}
+
+type labelsAndValue[T comparable] struct {
+	key    T
+	labels string // Prometheus-formatted {label="value",label="value"} string
+	val    expvar.Var
+}
+
+// labelString returns a Prometheus-formatted label string for the given key.
+func labelString(k any) string {
+	rv := reflect.ValueOf(k)
+	t := rv.Type()
+	if t.Kind() != reflect.Struct {
+		panic(fmt.Sprintf("MultiLabelMap must use keys of type struct; got %v", t))
+	}
+
+	var sb strings.Builder
+	sb.WriteString("{")
+
+	for i := 0; i < t.NumField(); i++ {
+		if i > 0 {
+			sb.WriteString(",")
+		}
+		ft := t.Field(i)
+		label := ft.Tag.Get("prom")
+		if label == "" {
+			label = strings.ToLower(ft.Name)
+		}
+		fv := rv.Field(i)
+		switch fv.Kind() {
+		case reflect.String:
+			fmt.Fprintf(&sb, "%s=%q", label, fv.String())
+		case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+			fmt.Fprintf(&sb, "%s=\"%d\"", label, fv.Int())
+		case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+			fmt.Fprintf(&sb, "%s=\"%d\"", label, fv.Uint())
+		case reflect.Bool:
+			fmt.Fprintf(&sb, "%s=\"%v\"", label, fv.Bool())
+		default:
+			panic(fmt.Sprintf("MultiLabelMap key field %q has unsupported type %v", ft.Name, fv.Type()))
+		}
+	}
+	sb.WriteString("}")
+	return sb.String()
+}
+
+// KeyValue represents a single entry in a [MultiLabelMap].
+type KeyValue[T comparable] struct {
+	Key   T
+	Value expvar.Var
+}
+
+func (v *MultiLabelMap[T]) String() string {
+	return `"MultiLabelMap"`
+}
+
+// WritePrometheus writes v to w in Prometheus exposition format.
+// The name argument is the metric name.
+func (v *MultiLabelMap[T]) WritePrometheus(w io.Writer, name string) {
+	if v.Type != "" {
+		io.WriteString(w, "# TYPE ")
+		io.WriteString(w, name)
+		io.WriteString(w, " ")
+		io.WriteString(w, v.Type)
+		io.WriteString(w, "\n")
+	}
+	if v.Help != "" {
+		io.WriteString(w, "# HELP ")
+		io.WriteString(w, name)
+		io.WriteString(w, " ")
+		io.WriteString(w, v.Help)
+		io.WriteString(w, "\n")
+	}
+	v.mu.RLock()
+	defer v.mu.RUnlock()
+
+	for _, kv := range v.sorted {
+		io.WriteString(w, name)
+		io.WriteString(w, kv.labels)
+		switch v := kv.val.(type) {
+		case *expvar.Int:
+			fmt.Fprintf(w, " %d\n", v.Value())
+		case *expvar.Float:
+			fmt.Fprintf(w, " %v\n", v.Value())
+		default:
+			fmt.Fprintf(w, " %s\n", kv.val)
+		}
+	}
+}
+
+// Init removes all keys from the map.
+func (v *MultiLabelMap[T]) Init() *MultiLabelMap[T] {
```

Comment:
Reset?

Reply:
I started by forking expvar.Map and it has https://pkg.go.dev/expvar#Map.Init 🤷‍♂️